### PR TITLE
Track type variable dependencies to guide instantiation decisions

### LINKED ISSUE #15864: Error on type inference
## Compiler version

3.1.3

## Minimized code

A similar issue (#11556) was already corrected several month ago (https://github.com/lampepfl/dotty/pull/12847)

```Scala
object Issue15864:
  type Traverser[-I, +O] = I => LazyList[(O)]
  extension[I, O](ta: Traverser[I, O])
    def ~[P](tb: Traverser[O, P]): Traverser[I, P] = ???

  class Graph { class Node; class Link }
  
  case class Path[+E](e: E)
  type Query[-U, +V] = Traverser[Path[U], Path[V]]

  def nodesQ(using g: Graph): Query[Nothing, g.Node] = ???
  def outsQ(using g: Graph): Query[g.Node, g.Node] = ???

  object aGraph extends Graph
  import aGraph._
  given aGraph.type = aGraph

  val q1: Query[Nothing, Node] = nodesQ ~ (outsQ ~ outsQ)
  implicitly[q1.type <:< Query[Nothing, Node]]

  val q2 = nodesQ ~ (outsQ ~ outsQ)
  implicitly[q2.type <:< Query[Nothing, Node]]

  val q3: Query[Nothing, Node] = q2
end Issue15864
```

## Output

```scala
Cannot prove that (asuivre.Issue15864.q2 : 
  asuivre.Issue15864.Traverser[asuivre.Issue15864.Path[Nothing], 
    asuivre.Issue15864.Path[Any]
  ]
) <:< asuivre.Issue15864.Query[Nothing, asuivre.Issue15864.aGraph.Node].
  implicitly[q2.type <:< Query[Nothing, Node]]
```

## Expectation
Supposed to compile without error.

### PR DESCRIPTION
**Pull request by odersky**
We now keep track of reverse type variable dependencies in constraints. 
E.g. if a constraint contains a clause like

    A >: List[B]

We associate with `B` info that `A` depends co-variantly on it. Or, if 

    A <: B => C

we associate with `B` that `A` depends co-variantly on it and with `C`
that `A` depends contra-variantly on it. Here, co-variant means that
the allowable range of `A` narrows if the referred-to variable `B` grows, and 
contra-variant means that the allowable range of `A` narrows if the referred-to 
variable `C` shrinks. If there's an invariant reference such as

    A <: Array[B]

Then `A` depends both co- and contra-variantly on `B`.

These dependencies are then used to guide type variable instantiation. 
If an eligible type variable does not appear in the type of a typed expression,
we interpolate it to one of its bounds. Previously this was done in an ad-hoc
manner where we minimized the type variable if it had a lower bound and maximized
it otherwise. We now take reverse dependencies into account. If maximizing a type
variable would narrow the remaining constraint we minimize, and if minimizing
a type variable would narrow the remaining constraint we maximize. Only if
the type variable is not referred to from the remaining constraint we resort
to the old heuristic based on the lower bound. 

Fixes #15864

Todo: This could be generalized in several directions:

 - We could base the dependency tracking on type param refs instead of type variables.
   That could make the `replace` operation in a constraint more efficient. 
 - We could base more interpolation decisions on dependencies. E.g. we could
   interpolate a type variable only if both the type of an expression and the enclosing
   constraint agree in which direction this should be done.

### CODE REVIEW

**Review by dwijnand — APPROVED** [review]

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
@mbovel @anatoliykmetyuk Benchmark bot seems to be down?

**Comment by mbovel** [comment]
test performance please

**Comment by odersky** [comment]
Some info on the parboiled2 test. It seems to be a hygiene violation caused somewhere in typedQuotePattern. The test failed when adding a new polytype to the constraint as follows (with -uniqid on, to disambiguate)
```
error while adding [T#1874900402]#1874900402
  (m#569565916: Map[String#6487, T#1874900402])
    (implicit h#1341031137: 
      org#15.parboiled2#2768.support#2770.HListable#2786[T#1874900402]
    ): org#15.parboiled2#2768.RuleN#2986[h#1341031137.Out#15632]/[T#1874900402] to #66778 with uninstantiated variables: T#1986710687, (param)1#489820775, 
  (param)2#1856085236
 constrained types: 
  [T#1986710687]#1986710687
    (m#620022507: Map[String#6487, T#1986710687], ignoreCase#620022507: 
      Boolean#88
    )
      (implicit h#1277453506: 
        org#15.parboiled2#2768.support#2770.HListable#2786[T#1986710687]
      ): org#15.parboiled2#2768.RuleN#2986[h#1277453506.Out#15632]
, 
  [(param)1#489820775 <: 
    org#15.parboiled2#2768.support#2770.HListable#2786[T#1986710687]
   & Singleton#1670]#489820775: h#1277453506.type
, 
  [(param)2#1856085236 <: 
    org#15.parboiled2#2768.support#2770.HListable#2786[T#1874900402]
   & Singleton#1670]#1856085236: h#1341031137.type
 bounds: 
     T#1986710687
     (param)1#489820775
     <: org#15.parboiled2#2768.support#2770.HListable#2786[T#1986710687] & 
      Singleton#1670
     (param)2#1856085236
     <: org#15.parboiled2#2768.support#2770.HListable#2786[T#1874900402] & 
      Singleton#1670
 ordering: 
 co-deps:
     T#1986710687 --> (param)1#489820775
 contra-deps:
     T#1986710687 --> (param)1#489820775
```
`T#1874900402` gets added as bound parameter to the constraint, but it is already referenced in the constraint. We then get:
```
java.lang.AssertionError: assertion failed: missing contra backwards reference T#1874900402 -> (param)2#1856085236 in #66779 with uninstantiated variables: T#1986710687, (param)1#489820775, 
  (param)2#1856085236
, T#1874900402
 constrained types: 
  [T#1986710687]#1986710687
    (m#620022507: Map[String#6487, T#1986710687], ignoreCase#620022507: 
      Boolean#88
    )
      (implicit h#1277453506: 
        org#15.parboiled2#2768.support#2770.HListable#2786[T#1986710687]
      ): org#15.parboiled2#2768.RuleN#2986[h#1277453506.Out#15632]
, 
  [(param)1#489820775 <: 
    org#15.parboiled2#2768.support#2770.HListable#2786[T#1986710687]
   & Singleton#1670]#489820775: h#1277453506.type
, 
  [(param)2#1856085236 <: 
    org#15.parboiled2#2768.support#2770.HListable#2786[T#1874900402]
   & Singleton#1670]#1856085236: h#1341031137.type
, 
  [T#1874900402]#1874900402
    (m#569565916: Map[String#6487, T#1874900402])
      (implicit h#1341031137: 
        org#15.parboiled2#2768.support#2770.HListable#2786[T#1874900402]
      ): org#15.parboiled2#2768.RuleN#2986[h#1341031137.Out#15632]
 bounds: 
     T#1986710687
     (param)1#489820775
     <: org#15.parboiled2#2768.support#2770.HListable#2786[T#1986710687] & 
      Singleton#1670
     (param)2#1856085236
     <: org#15.parboiled2#2768.support#2770.HListable#2786[T#1874900402] & 
      Singleton#1670
     T#1874900402
 ordering: 
 co-deps:
     T#1986710687 --> (param)1#489820775
 contra-deps:
     T#1986710687 --> (param)1#489820775
 while typechecking /Users/odersky/workspace/dotty/community-build/community-projects/parboiled2/parboiled-core/src/main/scala-3/org/parboiled2/support/OpTreeContext.scala
```

**Comment by odersky** [comment]
The perspective failure is also a hygiene violation (both added parameters are already referred to in the constraint). This time it's called from TypeComparer#matchCase.
```
error while adding [h#1295220417, t#1295220417 <: Tuple#52]#1295220417 =>> 
  scala#23.runtime#430.MatchCase#5264[h#1295220417 *: t#1295220417, h#1295220417
     
  | 
    scala#23.Tuple#53.Fold#23132[t#1295220417, Nothing#1128, 
      [x#-1965483054, y#-1965483054]#-1965483054 =>> x#-1965483054 | 
        y#-1965483054
    ]
  ]/[] to #23870 with uninstantiated variables: (param)1#673691532, (param)2#392956106, 
  (param)3#666882910
, B#1367453971, X0#1173905467
 constrained types: 
  [(param)1#673691532 <: 
    perspective#25.derivation#2791.HKDProductGeneric#11888[
      perspective#25.examples#2907.Cat#2932
    ]
   & Singleton#1676]#673691532: gen#1729214832.type
, 
  [(param)2#392956106 <: 
    perspective#25.derivation#2791.HKDProductGeneric#11888[
      perspective#25.examples#2907.Cat#2932
    ]
   & Singleton#1676]#392956106: gen#1729214832.type
, 
  [(param)3#666882910 <: 
    deriving#436.Mirror#11941.ProductOf#23108[
      perspective#25.examples#2907.Cat#2932
    ]{
      MirroredElemTypes = m$proxy1#40129.MirroredElemTypes; 
        MirroredLabel = m$proxy1#40129.MirroredLabel#23214
    }
   & Singleton#1676]#666882910: m#223342788.type
, 
  [B#1367453971 >: 
    m$proxy1#40129.MirroredElemLabels#23215 match {
      case EmptyTuple#12278 => Nothing#1128
      case h#1134074693 *: t#1134074693 => h#1134074693 | 
        scala#23.Tuple#53.Fold#23132[t#1134074693, Nothing#1128, 
          [x#-1965483054, y#-1965483054]#-1965483054 =>> x#-1965483054 | 
            y#-1965483054
        ]
    }
  ]#1367453971: Set[B#1367453971]
, [X0#1173905467]#1173905467: X0#1173905467
 bounds: 
     (param)1#673691532
     <: 
      perspective#25.derivation#2791.HKDProductGeneric#11888[
        perspective#25.examples#2907.Cat#2932
      ]
     & Singleton#1676
     (param)2#392956106
     <: 
      perspective#25.derivation#2791.HKDProductGeneric#11888[
        perspective#25.examples#2907.Cat#2932
      ]
     & Singleton#1676
     (param)3#666882910
     <: 
      deriving#436.Mirror#11941.ProductOf#23108[
        perspective#25.examples#2907.Cat#2932
      ]{
        MirroredElemTypes = m$proxy1#40129.MirroredElemTypes; 
          MirroredLabel = m$proxy1#40129.MirroredLabel#23214
      }
     & Singleton#1676
     B#1367453971
     >: 
      m$proxy1#40129.MirroredElemLabels#23215 match {
        case EmptyTuple#12278 => Nothing#1128
        case h#1295220417 *: t#1295220417 => h#1295220417 | 
          scala#23.Tuple#53.Fold#23132[t#1295220417, Nothing#1128, 
            [x#-1965483054, y#-1965483054]#-1965483054 =>> x#-1965483054 | 
              y#-1965483054
          ]
      }
     X0#1173905467
 ordering: 
 co-deps:
 contra-deps:
```

**Comment by odersky** [comment]
So, the question to decide is this:

 - We have evidence that reverse dependencies work except that there can be rare hygiene violations (2 cases in CB, none in test suite).
 - `replace` was not affected by these violations on the tests or the CB. It works based on dependencies or doing a brute force search of all parameters (which was the state before the last commit).
 - It's debatable what `replace` should do when faced with a hygiene problem. I.e say we have the constraint

        A >: C[B]

  then we add the binder [B] yielding

       A >: C[B]
       B

   and then we replace the bound `B` by `B1`. Should the pre-existing `C[B]` be replaced as well? Previously the answer was yes, with the replace optimization the answer is no.

So in the end the question is: Do we prefer to keep the behavior stable wrt replace or do we choose to optimize, with potentially big gains for large constraints?

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by mbovel** [comment]
test performance please

**Comment by mbovel** [comment]
test performance please

**Comment by mbovel** [comment]
test performance please

**Comment by odersky** [comment]
@smarter ping for review

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala [hunk 1]
```diff
@@ -455,24 +664,77 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
...
+            then
+              // That's the case if replace is called from TypeVar's instantiateWith.
+              // Forget about instantiation for old dependencies.
+              oldDepEntry = mapReplacedTypeVarTo(param)(oldDepEntry)
+            else
+              // That's the case if replace is called from unify.
+              // Assume parameter has been replaced for new dependencies
+              // (the actual replacement is done below).
+              newDepEntry = mapReplacedTypeVarTo(replacement)(newDepEntry)
+          case _ =>
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:706]: I'm having a hard time understanding what this does and what this code assumes exactly: would this need to be changed if the callers of `replace` changed? Maybe examples would help.

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala [hunk 2]
```diff
@@ -455,24 +664,77 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
...
+        if oldDepEntry ne newDepEntry then
+          if current eq this then
+            // We can end up here if oldEntry eq newEntry, so posssibly no new constraint
+            // was created, but oldDepEntry ne newDepEntry. In that case we must make
+            // sure we have a new constraint before updating dependencies.
+            current = newConstraint()
+          current.adjustDeps(newDepEntry, oldDepEntry, other)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:713]: Operations like `updateEntry` take `current` as an argument and check if it equals `this` to decide whether to copy or mutate the constraint, could `adjustDeps` do the same? The logic in replace is already very hard to follow so anything that can push complexity elsewhere would be welcome.

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala [hunk 3]
```diff
@@ -217,6 +240,188 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
...
+          seen += tp
+          traverse(tp.ref)
+      case _ => traverseChildren(t)
+  end Adjuster
+
+  /** Adjust dependencies to account for the delta of previous entry `prevEntry`
+   *  and the new bound `entry` for the type parameter `srcParam`.
+   */
+  def adjustDeps(entry: Type | Null, prevEntry: Type | Null, srcParam: TypeParamRef)(using Context): this.type =
+    val adjuster = new Adjuster(srcParam)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:346]: Because we reuse the same `Adjuster` instance for multiple traversals, `seen` might accumulate some entries in a call that might influence the next call. I would just create a new adjuster instance each time to avoid worrying about that.

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala [hunk 4]
```diff
@@ -217,6 +240,188 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
...
       else tvar
 
+// ------------- Type parameter dependencies ----------------------------------------
+
+  private type ReverseDeps = SimpleIdentityMap[TypeParamRef, SimpleIdentitySet[TypeParamRef]]
+
+  /** A map that associates type parameters of this constraint with all other type
+   *  parameters that refer to them in their bounds covariantly, such that, if the
+   *  type parameter is instantiated to a larger type, the constraint would be narrowed
+   *  (i.e. solution set changes other than simply being made larger).
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:250]: I'm not sure I understand what this means. What is being made larger?

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala [hunk 5]
```diff
@@ -217,6 +240,188 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
...
+
+  private class Adjuster(srcParam: TypeParamRef)(using Context)
+  extends TypeTraverser, ConstraintAwareTraversal[Unit]:
+
+    var add: Boolean = compiletime.uninitialized
+    private val seen = util.HashSet[LazyRef]()
+
+    def update(deps: ReverseDeps, referenced: TypeParamRef): ReverseDeps =
+      val prev = deps.at(referenced)
+      val newSet = if add then prev + srcParam else prev - srcParam
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:323]: Is this an optimization or does removing entries instead of doing nothing when `add == false` actually matter for correctness?

**Review by smarter — CHANGES_REQUESTED** [review]
First pass review, overall the part I'm least comfortable with is OrderingConstraint#replace which I haven't been able to wrap my mind around so far.
I'm also wondering if the information we're now keeping track of shouldn't let us simplify `Inferencing.variances`: do we still need to do a fixpoint iteration there?
It seems that we could also use this informatin to detect indirect cycles which currently crash the compiler: https://github.com/lampepfl/dotty/issues/15280

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala (hunk 4 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:250]: The solution set. i.e. the constraint is relaxed.

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:706]: I'll try to explain better.

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala (hunk 2 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:713]: I thought about it, but concluded that's not the same. `adjustDeps` is called to update internal data structures so that all internal invariants are satisfied. It does not create a new constraint, so there's generally no need to cache it. It's just in this specific corner case where it happens, and I tried to explain in the comment why.

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala (hunk 3 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:346]: I am going to `clear` the `seen` set instead. I am worried about creating too many adjusters otherwise.

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala (hunk 5 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:323]: Removing entries from `prev` could be necessary fro correctness. Otherwise we have stale backwards references to parameters that do not correspond to any forward reference. Whether this is a problem globally I am  not sure, but better not risk it. 

Removing the entry array from the deps set (in line 324) is a pure optimization. It also avoids printing junk.

**Review by smarter — CHANGES_REQUESTED** [review]
Otherwise I'm still wondering if coDeps/contraDeps couldreplace or help simplify the logic in [`Inferencing#variances`](https://github.com/lampepfl/dotty/blob/33818506801c80c8c73649fdaab3782c052580c6/compiler/src/dotty/tools/dotc/typer/Inferencing.scala#L446-L518).

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala [hunk 6]
```diff
@@ -217,6 +240,189 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
...
+   *   - When typing a prefx, don't avoid negative variances. This matters only for the
+   *     corner case where a parameter is instantiated to Nothing (see comment in
+   *     TypeAccumulator#applyToPrefix). When determining instantiation directions in
+   *     interpolations (which is what dependency variances are for), it can be ignored.
+   */
+  private trait ConstraintAwareTraversal[T] extends TypeAccumulator[T]:
+
+    override def tyconTypeParams(tp: AppliedType)(using Context): List[ParamInfo] =
+      def tparams(tycon: Type): List[ParamInfo] = tycon match
+        case tycon: TypeVar if !tycon.isInstantiated => tparams(tycon.origin)
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala]: Good point. I fixed it in https://github.com/lampepfl/dotty/pull/16042/commits/854c35dd6fed56b25f52d3fb914930e8edb49d46

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala [hunk 7]
```diff
@@ -217,6 +240,189 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
...
+  /** Null-safe indexing */
+  extension (deps: ReverseDeps) def at(param: TypeParamRef): SimpleIdentitySet[TypeParamRef] =
+    val result = deps(param)
+    if null == result  // swapped operand order important since `==` is overloaded in `SimpleIdentitySet`
+    then SimpleIdentitySet.empty
+    else result
+
+  override def dependsOn(tv: TypeVar, except: TypeVars, co: Boolean)(using Context): Boolean =
+    def origin(tv: TypeVar) =
+      assert(!tv.isInstantiated)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:270]:
```suggestion
      assert(!instType(tv).exists)
```

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala [hunk 8]
```diff
@@ -455,24 +665,81 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
...
+
+      def replaceParamIn(other: TypeParamRef) =
+        val oldEntry = current.entry(other)
+        val newEntry = current.ensureNonCyclic(other, oldEntry.substParam(param, replacement))
+        current = boundsLens.update(this, current, other, newEntry)
+        var oldDepEntry = oldEntry
+        var newDepEntry = newEntry
+        replacedTypeVar match
+          case tvar: TypeVar =>
+            if tvar.isInstantiated
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:697]:
```suggestion
            if tvar.inst.exists
```
(I believe we don't need to check `instType(tvar)` since at this point we haven't called updateEntry to instantiate the param in the constraint yet)

**Review by smarter — CHANGES_REQUESTED** [review]
There were two other uses of isInstantiated in OrderingConstraint which I suggest replacements for, otherwise LGTM.

**Comment by odersky** [comment]
@smarter The two occurrences were replaced following your suggestions.

File: compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala [hunk 9]
```diff
@@ -217,6 +240,189 @@ class OrderingConstraint(private val boundsMap: ParamBounds,
...
+// ------------- Type parameter dependencies ----------------------------------------
+
+  private type ReverseDeps = SimpleIdentityMap[TypeParamRef, SimpleIdentitySet[TypeParamRef]]
+
+  /** A map that associates type parameters of this constraint with all other type
+   *  parameters that refer to them in their bounds covariantly, such that, if the
+   *  type parameter is instantiated to a larger type, the constraint would be narrowed
+   *  (i.e. solution set changes other than simply being made larger).
+   */
+  private var coDeps: ReverseDeps = SimpleIdentityMap.empty
```
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:252]: I only just realised that this made OrderingConstraint into a mutable data structure (previously it was immutable).

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:252]: It's actually not mutable. `coDeps` and `contraDeps` are views that are maintained incrementally.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:252]: That would only make it not destructive.  And even so, there is a `dropDeps`, which does remove from them, if I'm not mistaken.

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:252]: I am not sure what you mean by destructive vs mutable. Are LazyLists mutable? In my book they are. And, `dropDeps` is not exported in `Constraint` so it is morally private (we should probably make it private to be clear about it)

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/core/OrderingConstraint.scala:252]: Ah, I understand better now.  You're right.

Even though it looks very different, I think I prefer the approach that's taken with the original components of OrderingConstraint, i.e. in ConstraintLens update, where depending on the previous constraint (which is the external, public reference) it was either mutated in place or copied on write, and then always returned the updated reference.  And that logic was well separate from the rest of the domain logic.  With these new variables, the updates just happen at some point of the logic, mutating the vars in place.
